packages/contracts-periphery/contracts/universal/faucet/Faucet.sol

```diff
@@ -0,0 +1,273 @@
+     *
+     * @param _name Contract name.
+     */
+    function initialize(string memory _name) public initializer {
+        __EIP712_init(_name, version());
+    }
+
+    /**
+     * @inheritdoc FaucetAuthModule
+     */
+    function verify(
+        Faucet.DripParameters memory params,
+        bytes memory id,
+        bytes memory proof
+    ) external view returns (bool) {
+        // Generate a EIP712 typed data hash to compare against the proof.
+        bytes32 digest = _hashTypedDataV4(
+            keccak256(
+                abi.encode(
+                    PROOF_TYPEHASH,
+                    params.recipient,
+                    params.nonce,
+                    id
+                )
+            )
+        );
+        return SignatureChecker.isValidSignatureNow(ADMIN, digest, proof);
+    }
+}
+
+/**
+ * @title  Faucet
+ * @notice Faucet contract that drips ETH to users.
+ */
+contract Faucet {
+    /**
+     * @notice Parameters for a drip.
+     */
+    struct DripParameters {
+        address payable recipient;
+        bytes32 nonce;
+    }
+
+    /**
+     * @notice Parameters for authentication.
+     */
+    struct AuthParameters {
+        FaucetAuthModule module;
+        bytes id;
+        bytes proof;
+    }
+
+    /**
+     * @notice Configuration for an authentication module.
+     */
+    struct ModuleConfig {
+        bool enabled;
+        uint256 ttl;
+        uint256 amount;
+    }
+
+    /**
+     * @notice Admin address that can configure the faucet.
+     */
+    address public immutable ADMIN;
+
+    /**
+     * @notice Mapping of authentication modules to their configurations.
+     */
+    mapping (FaucetAuthModule => ModuleConfig) public modules;
+
+    /**
+     * @notice Mapping of authentication IDs to the next timestamp at which they can be used.
+     */
+    mapping (FaucetAuthModule => mapping (bytes => uint256)) public timeouts;
+
+    /**
+     * @notice Maps from id to nonces to whether or not they have been used.
+     */
+    mapping(bytes => mapping(bytes32 => bool)) public usedNonces;
+
+	/**
+     * @notice Modifier that makes a function admin priviledged.
+     */
+    modifier priviledged() {
+        require(
+			msg.sender == ADMIN,
+			"Faucet: function can only be called by admin"
+        );
+		_;
+    }
+
+    /**
+     * @param admin Admin address that can configure the faucet.
+     */
+    constructor(
+        address admin
+    ) {
+        ADMIN = admin;
+    }
+
+    /**
+     * @notice Allows users to donate ETH to this contract.
+     */
+    receive() external payable {
+	    // Thank you!
+	}
+
+   /**
+     * @notice Allows the admin to withdraw funds.
+     *
+     * @param recipient Address to receive the funds.
+     * @param amount    Amount of ETH in wei to withdraw.
+     */
+    function withdraw(
+        address payable recipient,
+        uint256 amount
+    ) public priviledged {
+		new SafeSend{value: amount}(recipient);
```

Comment:
Why are you using `selfdestruct` to send funds? See the `SafeCall` library in `contracts-bedrock`. There is no guarantee that `selfdestruct` will work like this into the future as its scheduled to be removed from the protocol. It may or may not still send all of the funds depending on the particular EIP that is implemented

Reply:
Lawl that was me. Currently the most foolproof way of sending funds without triggering reentrancy.

Reply:
It seems highly likely that any future selfdestruct will still send funds, but fwiw it's easy enough to change this contract later.